src/IRM/IRMAdaptiveLinearKink.sol

```diff
@@ -0,0 +1,180 @@
+    /// @return The new interest rate at current utilization.
+    /// @return The new interest rate at kink.
+    function computeInterestRateInternal(address vault, uint256 cash, uint256 borrows)
+        internal
+        view
+        returns (uint256, int256)
+    {
+        // Calculate utilization rate.
+        uint256 totalAssets = cash + borrows;
+        int256 utilization = totalAssets == 0 ? int256(0) : int256(borrows) * WAD / int256(totalAssets);
+
+        // Calculate the normalized distance between current utilization and target utilization (kink).
+        // `err` is normalized to [-1,+1] where -1 is 0% util, 0 is at `kink` and +1 is 100% util.
+        int256 errNormFactor = utilization > kink ? WAD - kink : kink;
+        int256 err = (utilization - kink) * WAD / errNormFactor;
+
+        // Initialize rate if this is the first interaction.
+        IRState memory state = irState[vault];
+        if (state.lastUpdate == 0) return (calcLinearKinkRate(initialKinkRate, err), initialKinkRate);
+
+        // The speed is assumed constant between two updates, but it is in fact not constant because of interest.
+        // So the rate is always underestimated.
+        int256 speed = adjustmentSpeed * err / WAD;
+
+        // Calculate the adaptation parameter.
+        int256 deltaTime = int256(block.timestamp - state.lastUpdate);
+        int256 adaptation = speed * deltaTime;
+
+        // If adaptation == 0, avgKinkRate = endKinkRate = state.kinkRate.
+        if (adaptation == 0) return (calcLinearKinkRate(state.kinkRate, err), state.kinkRate);
+
+        // Formula of the average rate that should be returned:
+        // avg = 1/T * ∫_0^T linear_kink(state.kinkRate*exp(speed*x), err) dx
+        // The integral is approximated with the trapezoidal rule (N=2):
+        // avg ~= linear_kink([state.kinkRate + endKinkRate + 2*state.kinkRate*exp(speed*T/2)] / 4, err)
+        int256 endKinkRate = calcNewKinkRate(state.kinkRate, adaptation);
+        int256 midKinkRate = calcNewKinkRate(state.kinkRate, adaptation / 2);
+        int256 avgKinkRate = (state.kinkRate + endKinkRate + 2 * midKinkRate) / 4;
+
+        return (calcLinearKinkRate(avgKinkRate, err), endKinkRate);
+    }
+
+    /// @notice Calculate the interest rate according to the linear kink model.
+    /// @param kinkRate The current interest rate at kink.
+    /// @param err The distance between the current utilization rate and the kink, normalized to `[-1,1]`.
+    /// @dev rate = kinkRate * ((1 - 1/slope) * err + 1) if err < 0
+    ///             kinkRate * ((slope - 1) * err + 1) else.
+    /// @return The new interest rate at current utilization.
+    function calcLinearKinkRate(int256 kinkRate, int256 err) internal view returns (uint256) {
+        int256 coeff;
+        if (err < 0) {
+            coeff = WAD - WAD * WAD / slope;
+        } else {
+            coeff = slope - WAD;
+        }
+        return uint256(((coeff * err / WAD) + WAD) * kinkRate / WAD);
```

Comment:
EVK expects IR in ray, not wad

Reply:
this whole math I think has some assumptions made about the curve. In morpho, hardcoded minKinkRate = 1% , slope=4, and kink = 0.9 guarantees that the curve is always above 0, with absolute min IR = 0.25%, when utilization is 0 and kink rate is at minimum. But with freely configurable kink and minKinkRate, the curve could very well fall below 0. I don't think this code supports that at all, the IR can't be < 0. Anyway, I would consider moving the boundry values from the factory here and making sure they don't allow curves crossing the x axis. The factory constants guarantee that I think, the steepest curve before the kink is 1/1.01, so < 1 => with minRate of 10% it's ok.

Reply:
double checking - was 10% for minKinkRate selected on purpose or was it a typo? morpho has 1%

Reply:
minKinkRate: was a typo, fixed